### PR TITLE
fix: rewrite Adventure mappers, add codecov, other improvements

### LINKED ISSUE #186: Update Adventure mappers
### Confirmation

- [X] I have checked for similar issues.

### Problem

The current Adventure mappers do not support the latest version and should be generally improved.

### Suggested solution

The existing mappers should be completely replaced with a new, more efficient, and non-static system.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by joshuasing**
**Summary**
A lot of improvements but mostly related to Adventure mapping.
I need to learn how to focus on one thing before starting working on other things...

**Changes**
 - Separate publish and release workflow from gradle workflow.
 - Switch to `ncipollo/release-action@v1` for creating releases in publish workflow.
 - Update README.md with more information on platforms.
 - Reformat README.md.
 - annotations: Remove `Plugin.Platform`.
 - annotations: Rename `@PlatformDependeny` to `@Dependency`.
 - annotations: Fix bugs in generated main classes
 - Rewrite some Javadoc to make it easier to read.
 - Replace `PluginData` with `ChameleonPluginData` which has builders.
 - Rename plural packages to singular for consistency (users -> user, logging -> logger, etc.).
 - Add Preconditions to aid in checking states and validating arguments.
 - Rewrite Adventure mappers and audience reflection to adapt to Adventure changes and remove static use to avoid problems. (resolves #186)
 - Use `Collection` instead of `List` or `Set` in return types.
 - Add `EventSubscriber.Builder#filters(Collection)`.
 - Create `ServerEvent`.
 - Create `ConsoleUser`to represent platform consoles.
 - Add `Platform.<type>` constants, e.g. `Platform.BUKKIT -> "Bukkit"`.
 - Make PlatformTarget extend Predicate and a functional interface.
 - Add `PlatformTarget.<type>()` methods to match platforms by id, e.g. `PlatformTarget.bukkit()`.
 - Rename `ProxyUser#getServer()` to `ProxyUser#getConnectedServer()` for clarity.
 - Remove `User#chat(String)` and `User#chat(Component)` due to changes in Minecraft 1.19.1+.
 - Make `User` implement `net.kyori.adventure.identity.Identified`.
 - Rename `UserManager#getPlayers()` to `UserManager#getUsers()` for consistency.
 - Rename `UserManager#getPlayer(UUID)` to `UserManager#getUserById(UUID)` for consistency.
 - Add tests with 100% coverage for all Adventure mappers.
 - Add tests with 87% coverage for all Adventure audience reflection classes.
 - Improve format of `libs.versions.toml`.
 - Update adventure-platform to v4.2.0.
 - Update velocity-api to v3.3.1.
 - Update minestom to git:f291437ada.
 - Update annotations to v24.0.0.
 - Update errorprone to v2.18.0.
 - Update junit to v5.9.2.
 - Update nullaway to v0.10.8.
 - Improve nullaway configuration.
 - Add Codecov configuration and uploader.
 - Improve example project a little bit.
 - General reformatting and other improvements.
 - Bump version to `0.13.0-SNAPSHOT`.

**Checklist**
- [x] I acknowledge and agree to the terms of the [Developer Certificate of Origin](https://developercertificate.org/).
- [x] All contributed code can be distributed under the terms of the [MIT License](https://github.com/ChameleonFramework/Chameleon/blob/main/LICENSE).
- [x] I have read the [contributing guidelines](https://github.com/ChameleonFramework/Chameleon/blob/main/CONTRIBUTING.md) and [code of conduct](https://github.com/ChameleonFramework/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have checked the ["Allow edit from maintainers"](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) option.
- [x] I have added appropriate unit tests for my changes. 

**This pull request contains breaking changes.**

### CODE REVIEW

File: platform-bukkit/src/main/java/dev/hypera/chameleon/platform/bukkit/user/BukkitUser.java [hunk 1]
```diff
@@ -141,42 +139,26 @@ public boolean hasPermission(@NotNull String permission) {
...
-                return org.bukkit.GameMode.CREATIVE;
-            case ADVENTURE:
-                return org.bukkit.GameMode.ADVENTURE;
-            case SPECTATOR:
-                return org.bukkit.GameMode.SPECTATOR;
-            default:
-                return org.bukkit.GameMode.SURVIVAL;
-        }
+        Preconditions.checkNotNull("gameMode", gameMode);
+        this.player.setGameMode(org.bukkit.GameMode.valueOf(gameMode.name()));
```
Comment (LooFifteen) [inline platform-bukkit/src/main/java/dev/hypera/chameleon/platform/bukkit/user/BukkitUser.java:151]: To my knowledge, this method is much slower than a switch.

**Review by LooFifteen — APPROVED** [review]

File: api/src/main/java/dev/hypera/chameleon/Chameleon.java [hunk 2]
```diff
@@ -161,16 +170,25 @@ public void onDisable() {
...
         Constructor<?>[] platformExtensionConstructors = Arrays.stream(extensionAnnotation.value())
             .filter(p -> ChameleonUtil.getGenericTypeAsClass(p, 2).isAssignableFrom(getClass()))
-            .findFirst().map(Class::getConstructors).orElse(new Constructor<?>[0]);
+            .findFirst()
+            .map(Class::getConstructors)
+            .orElse(new Constructor<?>[0]);
 
-        if (platformExtensionConstructors.length < 1 || Arrays.stream(platformExtensionConstructors).noneMatch(c -> c.getParameterCount() == 0)) {
+        if (platformExtensionConstructors.length < 1 || Arrays.stream(platformExtensionConstructors)
+            .noneMatch(c -> c.getParameterCount() == 0)) {
```
[the remark this answers is not in the document]
Comment (joshuasing) [reply api/src/main/java/dev/hypera/chameleon/Chameleon.java:178]: That entire method should probably be rewritten or moved at some point, I may work on it as a part of majorly improving the extensions API.

File: api/src/main/java/dev/hypera/chameleon/command/SubCommand.java [hunk 3]
```diff
@@ -81,20 +84,25 @@ public final class SubCommand {
...
+        Preconditions.checkNotNull("context", context);
+        Preconditions.checkNotNull("parent", parent);
+
         try {
-            if (null != this.permission && !this.permission.value().isEmpty() && !context.getSender().hasPermission(this.permission.value())) {
-                parent.getPermissionErrorMessage().ifPresent(component -> context.getSender().sendMessage(component));
+            if (this.permission != null && !this.permission.value().isEmpty() &&
+                !context.getSender().hasPermission(this.permission.value())) {
+                parent.getPermissionErrorMessage().ifPresent(component ->
+                    context.getSender().sendMessage(component));
```
[the remark this answers is not in the document]
Comment (joshuasing) [reply api/src/main/java/dev/hypera/chameleon/command/SubCommand.java:105]: Yeah... will be replaced when the new command system is ready.

File: example/src/main/java/dev/hypera/chameleon/example/ChameleonExample.java [hunk 4]
```diff
@@ -21,86 +21,107 @@
...
 
+        // Custom event
+        chameleon.getEventBus().subscribe(ExampleCustomEvent.class, event -> {
+            this.logger.info("Received example custom event! Hello, " + event.getName() + "!");
+        });
+
+        /* Scheduling */
         chameleon.getScheduler().schedule(Task.builder(() -> {
             this.logger.info("This plugin has been running for 10 seconds!");
         }).delay(Schedule.seconds(10)).build());
```
[the remark this answers is not in the document]
Comment (joshuasing) [reply example/src/main/java/dev/hypera/chameleon/example/ChameleonExample.java]: It was

File: platform-bukkit/src/main/java/dev/hypera/chameleon/platform/bukkit/user/BukkitUser.java (hunk 1 above)
Comment (joshuasing) [reply platform-bukkit/src/main/java/dev/hypera/chameleon/platform/bukkit/user/BukkitUser.java]: Reverted :)

**Pull request merged**